lessons/advanced/typespec.md

```diff
@@ -0,0 +1,197 @@
+---
+layout: page
+title: Specifications and types
+category: advanced
+order: 9
+lang: en
+---
+
+If you want to define 'interface' like in Java or Ruby, Elixir contains syntax for that. More, if you want to define your custom type you can do that too. In this lesson we will learn about `@spec` and `@type` syntax.
+
+{% include toc.html %}
+
+## Introduction 
+
+It's not uncommon you would like to describe interface of your function. Of course You can use [@doc annotation](/lessons/basic/documentation), but it is only information for other developers that is not checked in compilation time. For this purpose Elixir has `@spec` annotation to describe specification of function that will be checked by compiler.
+
+However in some cases specification is going to be quite big and complicated. If you would like to reduce complexity, you want to introduce custom type definition. Elixir has `@type` annotation for that. In the other hand, Elixir is still dynamic language. That means all information about type will be ignored by compiler, but could be used by other tools.   
+
+## Specification
+
+If you have experience with Java or Ruby you could think about specification as an `interface`. Specification defines what should be type of function parameters and return value.
+
+To define input and output types we use `@spec` directive placed right before function definition and taking as a `params` name of function, list of parameter types, and after `::` type of return value.  
+
+Let's take a look at example:
+
+```elixir
+@spec sum_product(integer) :: integer
+def sum_product(a) do
+    [1, 2, 3]
+    |> Enum.map(fn el -> el * a end)
+    |> Enum.sum
+end
+```
+
+Everything looks ok and when we call valid result will be return, but function `Enum.sum` returns `number` not `integer` as we expected in `@spec`. It could be source of bugs!
```

Comment:
Maybe we could re-word this:

"Everything looks good and our new `sum_product/1` function returns a valid result, but is that all?  If we look at `Enum.sum/1` we'd find that it returns a `number` and not a `integer` as our `@spec` implies, this could be source of bugs!"